pkg/fqdn/cache.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fqdn
+
+import (
+	"bytes"
+	"net"
+	"sort"
+	"time"
+
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+var DefaultDNSCache = NewDNSCache()
+
+// cacheEntry objects are immutable once created.
+// They hold an the DNS data passed in with DNSCache.Update, nominally equating
+// to a DNS lookup. cacheEntry objects are managed by DNSCache and are not
+// expected to be returned.
+type cacheEntry struct {
+	Name           string
+	LookupTime     time.Time
+	ExpirationTime time.Time
+	TTL            int
+	IPs            []net.IP
+}
+
+func (entry *cacheEntry) isExpiredBy(pointInTime time.Time) bool {
+	return pointInTime.After(entry.ExpirationTime)
+}
+
+// ipEntries maps a unique IP to the cacheEntry that provides it in .IPs.
+// Multiple IPs may point to the same cacheEntry, or they may all be different.
+// Crucially, an IP may be present in a cacheEntry but the IP in ipEntries
+// points to another cacheEntry. This is because the second cacheEntry has a
+// later expiration for this specific IP, and may not include the other IPs
+// provided by the first entry.
+// The DNS name in the entries is not checked, but is assumed to be the same
+// for all entries.
+// Note: They are guarded by the DNSCache mutex.
+type ipEntries map[string]*cacheEntry
+
+// getIPs returns a sorted list of non-expired unique IPs.
+// This needs a read-lock
+func (s ipEntries) getIPs(now time.Time) []net.IP {
```

Comment:
Not sure how big of a concern this is - but say something holds the lock we need for reading here for a long time - wouldn't that allow entries to survive past the specified time?

Reply:
I'll assert that nothing should hold the lock for very long, but it will inevitably happen*. In the case you envision getIPs would use a time that is too old (from `Lookup`, before lock acquisition) and possibly see a different number of entries than if there was no lock delay. This is because the lock delay must have been due to an `Update` call that added new entries with later expiration times. It could see fewer because the Update may remove them in `removeExpired`.
I think your point is valid, though. A caller to `Lookup` likely wants the most valid data upon return, not upon function call. I've changed this to exactly what you suggested, by moving the locking to `Lookup`.
See my response to your Update/lookupTime comment. I'll make a related argument there.


* I assert this because the heaviest operation is Update and it is relatively quick. The numbers seem high to me but this is the worst case scenario where the cache is otherwise empty/unprimed.
```
for 1000 IPs per Update
PASS: cache_test.go:194: DNSCacheTestSuite.BenchmarkUpdateIPs	     200	   8850489 ns/op
```
```
for 10 IPs per Update
PASS: cache_test.go:194: DNSCacheTestSuite.BenchmarkUpdateIPs	  200000	      7646 ns/op
```

Reply:
Awesome, thank you for the in-depth reply!

Reply:
Ha, it was more for me. It all makes sense in my head... until it doesn't ;)